Dockerfile

```diff
@@ -1,15 +1,20 @@
-# iron/go:dev is the alpine image with the go tools added
-FROM golang
-WORKDIR /app
+# Build Container
+FROM golang:1.9.4-alpine3.7 AS build-env
+RUN apk add --no-cache --upgrade git openssh-client ca-certificates
+RUN go get -u github.com/golang/dep/cmd/dep
+WORKDIR /go/src/app
 
-# Set an env var that matches your github repo name
-ENV SRC_DIR=/go/src/github.com/Ice3man543/subfinder/
+# Cache the dependencies early
+COPY Gopkg.toml Gopkg.lock ./
+RUN dep ensure -vendor-only -v
 
-# Add the source code:
-ADD . $SRC_DIR
+# Build
+COPY *.go ./
```

Comment:
just a little detail, In docker documentation the supported syntax to allow correct layer caching and build seems to be listing files one by one or a directory.
This syntax seems to don't be aware of updates to main.go file, so the intermediate layer never gets updated. Changing it to:

```
COPY main.go ./
``` 
works for me on OSX and Linux.

+1 for using multi-stage builds